### PR TITLE
WP Stories integration - add Tracks (part 1)

### PR DESCRIPTION
**Pull request by mzorz**
This PR builds on top of  #12068 

This PR adds tracking for the WP Stories feature. Guided by looking into all the `// TODO WPSTORIES add TRACKS` comments to think and replace by proper behavior.

### Sequence of events
For a successful Story save and upload:
- trackEditorCreatedPost -> `EDITOR_CREATED_POST` this is when the StoryComposerActivity is shown for a new Story
- trackStorySaveResultEvent -> `STORY_SAVE_SUCCESSFUL`  (`STORY_SAVE_ERROR` if error)
- editorTracker.trackAddMediaFromDevice (`EDITOR_ADDED_VIDEO_VIA_DEVICE_LIBRARY` or `EDITOR_ADDED_PHOTO_VIA_DEVICE_LIBRARY`)
- trackSavePostAnalytics -> N/A (doesn't track anything because for a `PUBLISHED` Post it lets `PostUploadHandler` handle it)
- trackStoryPostSavedEvent -> `STORY_POST_SAVE_LOCALLY` if no network / `STORY_POST_SAVE_REMOTELY` in case it was correctly enqueued to be uploaded to the site. Uses `frame_qty` as a property to indicate the amount of frames this Story has.
- (PostUploadHandler) -> `EDITOR_PUBLISHED_POST` once it's correctly saved on the remote site.

### Changes in this PR
- de83be9 when a StoryResultEvent is received on the bridge, `trackStorySaveResultEvent` is called and tracks either `Stat.STORY_SAVE_SUCCESSFUL` or `Stat.STORY_SAVE_ERROR` with site details and the `isRetry` flag (boolean) to indicate whether this is a save retry by the user.
- 8ad12ec2bf adds STORY_POST_SAVE_LOCALLY and STORY_POST_SAVE_REMOTE events, which area used in the bridge as the last step (if there's network it's going to be saved remotely, if not, it's saved locally)
- d6de7a7317 tracks the specific case of touching the action label `MANAGE` on the error snackbar, instead of counting that as a error Notification tap via `ARG_NOTIFICATION_TYPE`.
-  d223dd4d0f9536b8715e3c636cacb4ec147904b8 moved `trackEditorCreatedPost` method from EditPostActivity to AnalyticsUtils so it's used both in EditPostActivity and StoryComposerActivity. This will track creation of new posts (while at it, also wrote tiny fix in https://github.com/wordpress-mobile/WordPress-Android/pull/12081 included in this PR as well)
- 29273bd includes the `post_format` property when adding postType property to tracks. This is of particular interest for Story posts given the only way to tell one another (from a regular Post) is by looking into this value (for Stories, it's `"wpstory"`)
- 4b1eaf4752 tracks the specific case for users tapping on the camera FAB in the media picker, signaling they want to open the StoryComposerActivity capture mode
- 0f9de40c43 removed `// TODO WPSTORIES add TRACKS` comment there given it's not needed: when picking media from the picker, tracks will be recorded in `trackAddMediaEvent`  in `AddExistingMediaToPostUseCase`.
- bae5282106b5eda13884c6c8780a19e9dcb9904d adds the use of `PostEditorAnalyticsSession` to StoryComposerActivity

#### Note
- for internal events while composing (emoji, text, etc.) we'll gather information in a subsequent PR. This PR tackles the WPAndroid integration side of things only.

#### To test:
- 4b1eaf4752 create a story, then tap on the Upload icon which opens the media picker, then tap on the camera button: observe the tracker gets called with value MEDIA_PICKER_OPEN_WP_STORIES_CAPTURE
- use all of the functionality, verify the code provided in the commits listed above is run

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java [hunk 1]
```diff
@@ -1114,11 +1112,7 @@ public void onEventMainThread(StorySaveEvents.StorySaveResult event) {
                                         event.getStoryIndex()) + ""
                         );
 
-                        // TODO WPSTORIES add TRACKS: the putExtra described here below for NOTIFICATION_TYPE
-                        // is meant to be used for tracking purposes. Use it!
-                        // TODO add NotificationType.MEDIA_SAVE_ERROR param later when integrating with WPAndroid
-                        //        val notificationType = NotificationType.MEDIA_SAVE_ERROR
-                        //        notificationIntent.putExtra(ARG_NOTIFICATION_TYPE, notificationType)
+                        intent.putExtra(ARG_NOTIFICATION_TYPE, NotificationType.STORY_SAVE_ERROR);
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java:1115]: minor note: this is related to the work done on #12068, but put it in this PR as I'm searching for all `// TODO WPSTORIES add TRACKS` comments and doing changes accordingly :)

File: WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerActivity.kt [hunk 2]
```diff
@@ -312,7 +312,7 @@ class StoryComposerActivity : ComposeLoopFrameActivity(),
         notificationIntent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
         notificationIntent.putExtra(WordPress.SITE, site)
         // setup tracks NotificationType for Notification tracking. Note this doesn't use our interface.
-        val notificationType = NotificationType.STORY_FRAME_SAVE_ERROR
+        val notificationType = NotificationType.STORY_SAVE_ERROR
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/stories/StoryComposerActivity.kt:315]: same here, related to #12068 - decided to use STORY_SAVE_ERROR as a more global error (will consider only keeping that one later)

File: WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java [hunk 3]
```diff
@@ -1114,11 +1114,10 @@ public void onEventMainThread(StorySaveEvents.StorySaveResult event) {
...
                         );
 
-                        // TODO WPSTORIES add TRACKS: the putExtra described here below for NOTIFICATION_TYPE
-                        // is meant to be used for tracking purposes. Use it!
-                        // TODO add NotificationType.MEDIA_SAVE_ERROR param later when integrating with WPAndroid
-                        //        val notificationType = NotificationType.MEDIA_SAVE_ERROR
-                        //        notificationIntent.putExtra(ARG_NOTIFICATION_TYPE, notificationType)
+                        mStoriesTrackerHelper.trackStorySaveResultEvent(
+                                event,
+                                Stat.STORY_SAVE_ERROR_SNACKBAR_MANAGE_TAPPED
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/main/MySiteFragment.java:1119]: ended up adding a new event for the specific case of tapping on the `manage` action on the error snackbar, given `NOTIFICATION_TYPE`  is reserved only for the case when the user taps on the error notification on the system dashboard

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 4]
```diff
@@ -441,7 +440,7 @@ protected void onCreate(Bundle savedInstanceState) {
         FragmentManager fragmentManager = getSupportFragmentManager();
         Bundle extras = getIntent().getExtras();
         String action = getIntent().getAction();
-        boolean isRestarting = !RestartEditorOptions.NO_RESTART.name().equals(extras.getString(EXTRA_RESTART_EDITOR));
+        boolean isRestarting = checkToRestart(getIntent());
```
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:443]: see https://github.com/wordpress-mobile/WordPress-Android/pull/12081 (unrelated to Stories, but needed to work well here)

**Comment by mzorz** [comment]
Thanks a ton for your review @aerych ! 

> It was interesting we track MEDIA_PICKER_OPEN_WP_STORIES_CAPTURE when tapping the camera button, but I didn't notice a similar camera capture event when entering capture mode initially. Probably no big deal if we're changing navigation but noting for future reference in case we decide differently and want to track the initial capture mode.

You're right - I think it's  mostly so we can track a conscious decision from the user to actually open the camera capture mode. And in the case we leave navigation as is, the actual screen being shown is all we need to identify the screen is being presented (that'd be `EDITOR_CREATED_POST `).

> It would be nice (at some point) if we could track when a user abandons the stories editor, and the number of frames they'd created before bailing out. Not important for the alpha but probably important to have for the beta / final release.

Yes! - I have plans for internal tracking (that's why this PR is named  "part 1"). The idea is we follow along the lines of `NotificationINtentLoader` and other various interfaces I've implemented here in order for the stories library to be able to tell the host app to track stuff on various points. I'll prepare a follow up PR with this later next week / the week after, that tracks various points that are internal to the actual Story creation experience, but need the host app to trigger an actual tracker 👍 

> I see where we're tracking the post_format as wpstory but I didn't see that reflected in the editor_post_published event in the tracks live view. There it was set to standard. I'm guessing this is most likely due to needing to whitelist wpstory as a valid post format on WPcom.

Oh interesting. Ok we'll need to look into that 🤔. Thanks for checking!

**Comment by oguzkocer** [comment]
Hi there! `15.2` is being frozen, so this PR is being bumped to `15.3`. If these changes _have to_ be shipped with `15.2`, please target `release/15.2` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java [hunk 5]
```diff
@@ -1479,6 +1479,8 @@ public static String getEventNameForStat(AnalyticsTracker.Stat stat) {
                 return "media_picker_device_library_opened";
             case MEDIA_PICKER_OPEN_WP_MEDIA:
                 return "media_picker_wordpress_library_opened";
+            case MEDIA_PICKER_OPEN_WP_STORIES_CAPTURE:
+                return "media_picker_wordpress_stories_capture_opened";
```
[the remark this answers is not in the document]
Comment (aforcier) [reply libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java:1517]: Agreed 👍

File: WordPress/src/main/java/org/wordpress/android/ui/stories/StoriesTrackerHelper.kt [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+class StoriesTrackerHelper @Inject constructor() {
+    fun trackStorySaveResultEvent(event: StorySaveResult) {
+        val stat = if (event.isSuccess()) Stat.STORY_SAVE_SUCCESSFUL else Stat.STORY_SAVE_ERROR
+        trackStorySaveResultEvent(event, stat)
+    }
+
+    fun trackStoryPostSavedEvent(frameQty: Int, site: SiteModel, locallySaved: Boolean) {
+        val stat = if (locallySaved) Stat.STORY_POST_SAVE_LOCALLY else Stat.STORY_POST_SAVE_REMOTELY
+        val properties: HashMap<String, Any> = HashMap()
+        properties.put("frame_qty", frameQty)
```
[the remark this answers is not in the document]
Comment (mzorz) [reply WordPress/src/main/java/org/wordpress/android/ui/stories/StoriesTrackerHelper.kt]: I changed it in 5f99004e68 but also we should remove all references to `frame` in for example `FrameSaveService` and so on. We don't need to do that now

File: libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (mzorz) [reply libs/analytics/WordPressAnalytics/src/main/java/org/wordpress/android/analytics/AnalyticsTrackerNosara.java:1517]: okay... ef26b0037c

**Comment by mzorz** [comment]
Thanks for your review and thorough testing @aforcier ! 🙇 

> It may not make sense to fire those events earlier, but it seems to me we still need some way to mark 'user chose "Add new story"', since this would be the start of the story creation flow and if they backed away from the media picker that should still count as 'started story creation flow'.

👍 agreed, will create a new PR for this on top of the stack (given we need to add this on top of the new navigation).

> I was a bit surprised that the editor_post_published event has "editor_source":"gutenberg" - I expected to find wp_stories_creator. It looks like the allowed values for editor_source are gutenberg or aztec. This seems a bit inconsistent to me and maybe different from what you intended - I don't think it's a deal breaker but wanted to point it out.

Nice catch; the specific place where this came from is actually here https://github.com/wordpress-mobile/WordPress-Android/blob/aa801c05a4063fc39167d5ae70f1726a5bef622a/WordPress/src/main/java/org/wordpress/android/ui/uploads/PostUploadHandler.java#L330-L340, but you're right there's only one other place where this is used and that is in PostUtils as you pointed out, but that's for `SCHEDULED` posts only (not our case FWIW).

I'll address this on a subsequent PR on top of the stack, given I'd need to check whether the post contains any Story blocks to identify the `editor_source` correctly 👍 

> Finally, I'm a bit concerned about using a post format to distinguish story posts. We need it (or something like it) as a way to recognize story posts, and I'm looking into some options there - but I wonder if we can avoid relying on it for event tracking and have a separate, unambiguous event for 'story_created' that fires when editor_post_published does. What do you think about that @mzorz?

That's fair. I think we need a way to identify these posts even when there's still a possibility of ambiguity or false positives for things like for example, showing on the Reader or providing a different view in Posts lists. As discussed elsewhere witth @aerych,  (just adding here for completeness) such an ambiguity would only exist if the format was changed manually by the user (sounds rare), or if they'd edited the Post content manually and removed any Story blocks (seems potentially more likely).

One thing to note: at the time of the creation there won't be any ambiguity: when `editor_post_published` is triggered it will come with the post format property correctly set, and even if the user may change it later, the tracked information makes it univocally distinguishable that this event belongs to a post created with the Story creator. If we still would like to have a separate event (for reasons such as it'd be easier to make queries afterwards or any other reason) we can still add it, but just wanted to make sure this was clear.

**Comment by aforcier** [comment]
> That's fair. I think we need a way to identify these posts even when there's still a possibility of ambiguity or false positives for things like for example, showing on the Reader or providing a different view in Posts lists. As discussed elsewhere witth @aerych, (just adding here for completeness) such an ambiguity would only exist if the format was changed manually by the user (sounds rare), or if they'd edited the Post content manually and removed any Story blocks (seems potentially more likely).

That's correct, but the situation might be worse than that - as @aerych pointed out the new post format hasn't been allowlisted so we it doesn't look like we can currently set a post to have that format. And that's where my hesitation is coming from - AFAIK post formats are altered by themes, I'm not certain the new custom post format existing would be e.g. something we could rely on for Jetpack, or later on non-Jetpack self-hosted sites. Something to figure out in the coming weeks.

> One thing to note: at the time of the creation there won't be any ambiguity: when editor_post_published is triggered it will come with the post format property correctly set, and even if the user may change it later, the tracked information makes it univocally distinguishable that this event belongs to a post created with the Story creator. If we still would like to have a separate event (for reasons such as it'd be easier to make queries afterwards or any other reason) we can still add it, but just wanted to make sure this was clear.

This can be true eventually, though at the moment it doesn't look from my tests that the correct format is even sent to the server, so the tracking part doesn't work yet either:
```
I/WordPress-STATS: 🔵 Tracked: editor_post_published, Properties: {"editor_source":"gutenberg","word_count":1,"blog_id":11111111,"post_id":91,"has_gutenberg_blocks":true,"post_type":"post","post_format":"standard","is_jetpack":true}
```

Does it work for you? If it's not just on my end, my guess without digging in further just yet is that we're checking the story post format against the supported post formats for that site from the API and are defaulting to the site's default since the story format is not in the supported list.

I think there might be an argument to be made that it might be poor data health to tag posts as having a post format that they don't end up having - but we can visit that once we have an answer to how far we can rely on post format in general.

**Comment by mzorz** [comment]
> as @aerych pointed out the new post format hasn't been allowlisted so we it doesn't look like we can currently set a post to have that format. And that's where my hesitation is coming from - AFAIK post formats are altered by themes, I'm not certain the new custom post format existing would be e.g. something we could rely on for Jetpack, or later on non-Jetpack self-hosted sites. Something to figure out in the coming weeks.

Another thing we could do, and this has been discussed as well is to lurk into the each post's content and see if a story block is present. We're doing that already in the apps to check whether a Post contains Gutenberg blocks for instance, but I'm not happy with doing that on the fly for posts lists, given it could take long for longer posts that may not even have Gutenberg blocks at all. Maybe we could preprocess or only take up to the first n bytes to evaluate, knowing we could miss some actual positives, but this would at.least not make the entire experience clunky when scrolling through a list.
As you can see this also has the same problem in terms of not being able to.identify a post as containing story blocks correctly, so if I had to decide I think the format may still be a good candidate solution to evaluate. We.could probably draw some data from tracks and evaluate which of the two potential solutions appears to have better results on a given data set? 


> This can be true eventually, though at the moment it doesn't look from my tests that the correct format is even sent to the server, so the tracking part doesn't work yet either:

I was meaning to refer to this PR which fixes those parts that we agreed on ;)
https://github.com/wordpress-mobile/WordPress-Android/pull/12345
Can you check if it fixes that particular issue for you @aforcier?

**Comment by oguzkocer** [comment]
Hi there! `15.3` is being frozen, so this PR is being bumped to `15.4`. If these changes _have to_ be shipped with `15.3`, please target `release/15.3` and cc me in the PR so I can cut a new beta release. Also if you can keep the release dates in mind and bump the PRs & issues you are working on before the code freeze date, we'd really appreciate it.

_This is a standard release version bump message. Let me know if you have any feedback._

**Comment by oguzkocer** [comment]
On second thought, we no longer need to set the milestone for PRs that don't target `develop` or `release` branches. Milestones should be set to whichever version of the app the feature will be released in, so the milestones for PRs against feature branches are unlikely to be correct. We just need to make sure the PR from the feature branch to `develop` has the milestone which Peril checks will ensure. So, I am removing the milestone for this PR.

**Comment by mzorz** [comment]
Heads up:  

**tl;dr** make sure to test code directly on #12345 branch ([`feature/wp-stories-tracks-2`](https://github.com/wordpress-mobile/WordPress-Android/tree/feature/wp-stories-tracks-2), which includes the changes in this very PR #12074).

*Long story:*
#12345 which fixes the problems mentioned in the comments by @aforcier up here above has been merged in the merge-chain for some involuntary reason - after this  one here https://github.com/wordpress-mobile/WordPress-Android/pull/12306 was updated  from targeting feature/wp-stories/alpha1 to targeting `feature/wp-stories/alpha2`  I "safely" deleted the branch for https://github.com/wordpress-mobile/WordPress-Android/pull/12333 (`issue-12330/bottom_sheet_publish_flow`) that was targeted to branch `feature/wp_stories_bottomsheet` when I did that, this one https://github.com/wordpress-mobile/WordPress-Android/pull/12345 got automagically merged (branch feature/wp-stories-tracks-2 targeting same base branch as that other PR, `feature/wp_stories_bottomsheet`).
Given it was easier to update the remaining branches to clear the mess, I did that now and I'm making this comment here so when the reviewer (either @jd-alexander or @aforcier) checks back on this make sure to test code directly on that branch ([`feature/wp-stories-tracks-2`](https://github.com/wordpress-mobile/WordPress-Android/tree/feature/wp-stories-tracks-2), which includes the changes in this very PR #12074).

**Comment by mzorz** [comment]
> Hi @mzorz thanks for these changes. I did indeed test them on the #12345 branch to verify that the tracked events worked as expected and they did.

Thank you for reviewing @jd-alexander 🙇 

> The only unrelated issue I found was when I added a video to the story it was detected as a picture in the story composer and exported as such. I am looking into it but sharing here for clarity.

This is interesting, can you add an issue so we make sure to track it down? 🙏 thanks in advance.

Merging this one 👍

**Comment by jd-alexander** [comment]
> This is interesting, can you add an issue so we make sure to track it down? 🙏 thanks in advance.

It seems as if this was fixed with the latest changes because I can't reproduce it in the `alpha2` branch so we are good 🙏
